paddle/fluid/framework/details/ssa_graph_builder.cc

```diff
@@ -37,49 +37,56 @@ void SSAGraphBuilder::PolishGraphToSupportDataHazards(SSAGraph *graph) {
             continue;
           }
 
-          auto *dep_var = new DummyVarHandle();
+          auto *dep_var = new DummyVarHandle(graph->CreateEmptyNode("dummy"));
           read_op->AddOutput(dep_var);
           write_op->AddInput(dep_var);
-          graph->dep_vars_.emplace(dep_var);
+          graph->Get<GraphDepVars>("dep_vars").emplace(dep_var);
         }
       }
     }
   }
 }
 
 VarHandle *SSAGraphBuilder::CreateOrGetLatestVarHandle(
-    SSAGraph *graph, const std::string &each_var_name,
-    const platform::Place &place, size_t place_offset) {
-  auto &var_holders = graph->vars_[place_offset];
-  auto &var_holder = var_holders[each_var_name];
+    Graph *graph, ir::Node *node, const platform::Place &place,
+    size_t place_offset) {
+  auto &var_holders = graph->Get<GraphVars>("vars")[place_offset];
+  auto &var_holder = var_holders[node->Name()];
   VarHandle *var = nullptr;
   if (var_holder.empty()) {
-    var = new VarHandle(0, place_offset, each_var_name, place);
+    if (node->NodeType() == ir::Node::Type::kVariable) {
+      var = new VarHandle(graph->CreateVarNode(node->Var()), 0, place_offset,
+                          node->Name(), place);
+    } else {
+      var = new VarHandle(graph->CreateEmptyNode(node->Name()), 0, place_offset,
+                          node->Name(), place);
+    }
     var_holder.emplace_back(var);
   } else {
     var = var_holder.rbegin()->get();
   }
   return var;
 }
 
-void SSAGraphBuilder::CreateOpOutput(SSAGraph *graph, OpHandleBase *op_handle,
-                                     const std::string &each_var_name,
+void SSAGraphBuilder::CreateOpOutput(Graph *graph, OpHandleBase *op_handle,
+                                     ir::Node *node,
                                      const platform::Place &place,
                                      size_t place_offset) {
-  auto &vars = graph->vars_[place_offset][each_var_name];
+  auto &vars = graph->Get<GraphVars>("vars")[place_offset][node->Name()];
```

Comment:
Is this a temporary solution? SSAGraph should `is a` `Graph` not a `Graph.attrs`

Reply:
The general principle is to put things shared by all passes as Graph class member. Other information that is specific to some passes will be kept as attrs.

Variables with versions is likely be part of Graph class in the future. But it might not be a separate field. Var is just a type of node. Hence it will likely be in the list of ```Graph.nodes```.

Pass writers are free to build collections like ```GraphVar``` here. It's just helper collections that make some lookup and insertions easier. It points to the nodes in ```Graph.nodes```.

I have a separate PR that moves SSA Graph building into a standalone method.
https://github.com/PaddlePaddle/Paddle/pull/12169